### ISSUE TITLE
Tutorial Application

### ISSUE THREAD

**Issue by pharringtonp19** [jacobjinkelly/easy-neural-ode#5]
It would be great to see how this regularized training method can be applied to [the Implicit Layers Neural Ode Tutorial](http://implicit-layers-tutorial.org/neural_odes/).

Just to increase one's understanding, it would be helpful to see (1) how the regularization fits into the training loop and (2) how jet differs from applying first order A.D. repeatedly.

I am going to try and tackle this over the next week, but would appreciate any help from anyone who may be interested! -- Thanks

**Comment by pharringtonp19** [comment]
Initial Results look promising -- the code is pretty easy to follow as is -- still trying to better understand how `jet` works together with `sol_recursive(f, z, t)`
![neural_ode_regularization](https://user-images.githubusercontent.com/55798098/124963153-777f3280-dfed-11eb-8ac6-c2a11ef79350.png)

**Comment by jacobjinkelly** [comment]
Hi @pharringtonp19, these are really cool results! I'm glad to see that you've found this repo useful. To answer your questions:

`sol_recursive` is essentially a special-case of `jet` applied to `odeint`.  Ideally we would have explicitly implemented `sol_recursive` directly as part of the `jet` API, but we found it easier in terms of engineering to keep it separate. You can find some more details of the math in A.2 of the [paper](https://arxiv.org/abs/2007.04504) and Section 4 in the earlier [workshop paper](https://openreview.net/forum?id=SkxEF3FNPH).

(1) The regularization is added to the loss [here](https://github.com/jacobjinkelly/easy-neural-ode/blob/master/mnist.py#L467). The regularization dynamics themselves are defined [here](https://github.com/jacobjinkelly/easy-neural-ode/blob/master/mnist.py#L291).

(2) As an example, you can find the jet rule for the `log` primitive [here](https://github.com/google/jax/blob/e9611eb0904409f3c1a5b0e585fa7d91c8721dcf/jax/experimental/jet.py#L430). `v` has length K equal to the order of derivative we want to compute, and you can see that there's a nested for loop, making the time complexity O(K^2). In contrast, repeatedly nesting `grad` (first-order AD) would result in exponential time. Essentially this is because `grad(f)` costs 3x to evaluate compared to the cost of f (we need to do a forward pass on `f`, which 1x, and backward pass on `f` which is 2x). Thus `grad(grad(f))` is approximately 9x the cost of `f`, and `grad` nested K times is going to be exponential in K. Essentially, `jet` is more efficient since it works directly with the Taylor coefficients for each primitive and avoids doing repeat work while `grad` doesn't.

Let me know if anything is unclear or there's something I can explain more thoroughly, or if you have any other questions. I'm happy to help :D

**Comment by pharringtonp19** [comment]
@jacobjinkelly Thanks for the explanation. I need to dig deeper into this.

Would it be possible to include an additional regularization term that penalizes how much dy/dt varies with y?

**Comment by pharringtonp19** [comment]
@jacobjinkelly I don't know if you would have any thoughts on the following problem -- I would like to extend [Will Whitney's](http://willwhitney.com/parallel-training-jax.html) approach of training multiple feedforward neural networks on a single gpu  to case of training multiple neural odes (via your regularization technique) on a single gpu.

Will's approach works as follows: for each training epoch, he vmaps the gradient update step across a set of networks. 

I don't think this is nearly as efficient when training a neural ode via an adaptive step size solver (but I may be wrong!) because the duration of each gradient update can vary across networks. Therefore, if we vmap the gradient update step across a set of networks, for each network, we cannot proceed onto the next training epoch until all networks have finished updating. This seems slow, and in my initial attempts at training neural odes in the fashion, it is.

 Given your experience with neural odes and Jax, I would be curious to know if you had any thoughts. -- thanks

**Comment by pharringtonp19** [comment]
@jacobjinkelly Feel free to disregard this -- it probably requires more thinking on my end -- I am wondering if, based on the implementation, it is possible that this type of regularizing  can actually increase the `nfe` count

At a high level, if we want to integrate something besides the inputs, we can simple "tack" that thing onto our "state vector": `x(t)` where `x(0)` are the inputs at time zero. As in your paper, if we want to integrate a higher order derivative of the dynamics function w.r.t time, we can add an extra dimension to our input vector (initialized at zero), `[x(0), 0]`, and then in an "altered dynamics function" compute both the Dynamics and the high order derivative together. 

Given that we're using an adaptive solver, is it possible that by adding this derivative dimension, we could increase the solver time? Would it make sense to allow the error tolerance levels for our solver to differ across the dimensions of our state vector? That is, for the transformed inputs, `x(t)`, we would want low tolerance, but for the higher order derivatives, we would allow for a higher tolerance.

I admit, I am not as familiar with your implementation/the implementation in jax as I should be, so perhaps you side step this issue in your implementation in such a way that the adaptive solver is not even exposed to the part of the state that corresponds to the higher order derivative. 

-- I would be curious to hear your thoughts. Thanks again for your time

**Comment by jacobjinkelly** [comment]
@pharringtonp19 Thanks for your question, and my apologies for the delayed reply. You're right - our regularization does increase the `nfe` count, for the reasons you've stated. However, at test-time, we don't need to compute the regularization since we only care about the actual predictions of the model. Thus, while training might incur higher `nfe`, at test-time we can get lower `nfe`. This is a tradeoff of course, and we discussed this tradeoff more in the paper. The other thing I'll mention is that we actually only need to integrate the regularization term on the backwards pass - on the forwards pass we can just integrate a placeholder value of `0`. This is essentially because we only care about the gradient with respect to the regularization term, and the gradient of the regularization can be constructed from the value of the rest of the state (everything but the regularization) and its gradient.

We wanted to make our method as solver agnostic as possible, so we didn't investigate modifying the tolerance levels across the dimension. You can check out [this paper](https://arxiv.org/abs/2009.09457) which explores ideas related to this, and [this one](https://arxiv.org/abs/2105.03918) too.

**Comment by pharringtonp19** [comment]
@jacobjinkelly I am going to see if it's possible to implement your method with [diffrax](https://docs.kidger.site/diffrax/usage/getting-started/). Thanks for all your pointers regarding my naive implementation!
